### PR TITLE
Million-AID dataset

### PR DESCRIPTION
**Pull request by nilsleh**
This PR adds the [Million-AID](https://captain-whu.github.io/DiRS/) dataset which contains one million aerial scenes from Google Earth engine. 

Comments/Questions:
- It offers both a multi-class (51 classes) and a multi-label (73 labels) task, for which I added an option in the constructor
- images can have two or three labels in the multi-label case so the `__getitem__` currently returns variable length label tensor
- I am not sure how to best handle the download because the total file is more than 200GB, so is there a way to maybe split it somehow?

Plot Examples:
<p float="left">

### CODE REVIEW

File: torchgeo/datasets/millionaid.py [hunk 1]
```diff
@@ -0,0 +1,397 @@
...
+        """
+        imgs_no_subcat = list(
+            glob.glob(os.path.join(root, self.split, "**", "**", "*.jpg"))
+        )
+
+        imgs_subcat = list(
+            glob.glob(os.path.join(root, self.split, "**", "**", "**", "*.jpg"))
+        )
+
+        scenes = [p.split("/")[-3] for p in imgs_no_subcat] + [
```
Comment (adamjstewart) [inline torchgeo/datasets/millionaid.py:277]:
```suggestion
        scenes = [p.split(os.sep)[-3] for p in imgs_no_subcat] + [
```
Same for other instances. This will fix the failing Windows tests.

**Comment by adamjstewart** [comment]
Bleh, I'll fix the mypy issues in another PR, it looks like PyTorch finally added type hints for many of its functions in the latest release that came out yesterday.

**Comment by calebrob6** [comment]
@nilsleh can you rebase on main here

**Comment by adamjstewart** [comment]
Rebasing isn't necessary, just need to remove the few remaining type ignores that are causing the tests to fail.

File: tests/data/millionaid/data.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
+#!/usr/bin/env python3
+
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+import hashlib
+import os
+import random
```
Comment (adamjstewart) [inline tests/data/millionaid/data.py:8]: `random` isn't used in this file, this can be removed

File: tests/datasets/test_millionaid.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+class TestMillionAID:
+    @pytest.fixture(
+        params=zip(
+            ["train", "train", "test", "test"],
+            ["multi-class", "multi-label", "multi-class", "multi-label"],
+        )
+    )
+    def dataset(
+        self,
+        monkeypatch: Generator[MonkeyPatch, None, None],
```
Comment (adamjstewart) [inline tests/datasets/test_millionaid.py:33]:
```suggestion
        monkeypatch: MonkeyPatch,
```

File: tests/datasets/test_millionaid.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+def download_url(url: str, root: str, *args: str) -> None:
+    shutil.copy(url, root)
+
+
+class TestMillionAID:
+    @pytest.fixture(
+        params=zip(
+            ["train", "train", "test", "test"],
+            ["multi-class", "multi-label", "multi-class", "multi-label"],
+        )
```
Comment (adamjstewart) [inline tests/datasets/test_millionaid.py:29]: You could use `itertools.product`, or just `zip` and only run two tests

File: torchgeo/datasets/millionaid.py [hunk 5]
```diff
@@ -0,0 +1,397 @@
...
+from torchgeo.datasets import VisionDataset
+
+from .utils import check_integrity, download_and_extract_archive, extract_archive
+
+
+class MillionAID(VisionDataset):
+    """Million-AID Dataset.
+
+    The `MillionAID <https://captain-whu.github.io/DiRS/>`_ dataset consists
+    of one million aerial images from the Googl Earth Engine that offers
```
Comment (adamjstewart) [inline torchgeo/datasets/millionaid.py:24]:
```suggestion
    of one million aerial images from Google Earth Engine that offers
```

File: torchgeo/datasets/millionaid.py [hunk 6]
```diff
@@ -0,0 +1,397 @@
...
+            labels = [
+                [sc, sub, c] if sub != "Missing" else [sc, c]
+                for sc, sub, c in zip(scenes, subcategories, classes)
+            ]
+        else:
+            labels = [[c] for c in classes]
+
+        images = imgs_no_subcat + imgs_subcat
+
+        files = [dict(image=img, label=l) for img, l in zip(images, labels)]
```
Comment (adamjstewart) [inline torchgeo/datasets/millionaid.py:299]: `pyupgrade` (a new CI step we added recently) will probably complain about dict comprehension here. You can run `pip install pyupgrade` and `pyupgrade --py37-plus torchgeo/datasets/millionaid.py` to fix this.

File: torchgeo/datasets/millionaid.py [hunk 7]
```diff
@@ -0,0 +1,397 @@
...
+        if os.path.isfile(filepath):
+            if self.checksum and not check_integrity(filepath, self.md5s[self.split]):
+                raise RuntimeError("Dataset found, but corrupted.")
+            extract_archive(filepath)
+            return
+
+        # Check if the user requested to download the dataset
+        if not self.download:
+            raise RuntimeError(
+                "Dataset not found in `root` directory, either specify a different"
```
Comment (adamjstewart) [inline torchgeo/datasets/millionaid.py:339]: We should print the root directory so it's more clear where we are looking. See CDL for an example of this.

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: torchgeo/datasets/millionaid.py [hunk 8]
```diff
@@ -0,0 +1,397 @@
...
+from torchgeo.datasets import VisionDataset
+
+from .utils import check_integrity, download_and_extract_archive, extract_archive
+
+
+class MillionAID(VisionDataset):
+    """Million-AID Dataset.
+
+    The `MillionAID <https://captain-whu.github.io/DiRS/>`_ dataset consists
+    of one million aerial images from the Google Earth Engine that offers
```
Comment (adamjstewart) [inline torchgeo/datasets/millionaid.py:24]:
```suggestion
    of one million aerial images from Google Earth Engine that offers
```
GEE is a proper noun, no need for "the"

File: torchgeo/datasets/millionaid.py [hunk 9]
```diff
@@ -0,0 +1,397 @@
...
+    either `a mult-class learning task
```
Comment (adamjstewart) [inline torchgeo/datasets/millionaid.py:25]:
```suggestion
    either `a multi-class learning task
```

File: torchgeo/datasets/millionaid.py [hunk 10]
```diff
@@ -0,0 +1,397 @@
...
+    <https://competitions.codalab.org/competitions/35945#learn_the_details-dataset>`_
+    with 51 classess or a `mulit-label learning task
```
Comment (adamjstewart) [inline torchgeo/datasets/millionaid.py:27]:
```suggestion
    with 51 classes or a `multi-label learning task
```

File: torchgeo/datasets/millionaid.py [hunk 11]
```diff
@@ -0,0 +1,397 @@
...
+        if self.transforms is not None:
+            sample = self.transforms(sample)
+
+        return sample
+
+    def _load_files(self, root: str) -> List[Dict[str, Any]]:
+        """Return the paths of the files in the dataset.
+
+        Args:
+            root: root dir of dataset
```
Comment (adamjstewart) [inline torchgeo/datasets/millionaid.py:264]:
```suggestion
            root: root directory of dataset
```

File: torchgeo/datasets/millionaid.py [hunk 12]
```diff
@@ -0,0 +1,397 @@
...
+
+        Returns:
+            list of dicts containing paths for each pair of image, and list of labels
+        """
+        imgs_no_subcat = list(
+            glob.glob(os.path.join(root, self.split, "**", "**", "*.jpg"))
+        )
+
+        imgs_subcat = list(
+            glob.glob(os.path.join(root, self.split, "**", "**", "**", "*.jpg"))
```
Comment (adamjstewart) [inline torchgeo/datasets/millionaid.py:274]: Do you mean:
```suggestion
            glob.glob(os.path.join(root, self.split, "*", "*", "*.jpg"))
        )

        imgs_subcat = list(
            glob.glob(os.path.join(root, self.split, "*", "*", "*", "*.jpg"))
```
or:
```suggestion
            glob.glob(os.path.join(root, self.split, "**", "*.jpg"))
        )

        imgs_subcat = list(
            glob.glob(os.path.join(root, self.split, "**", "*.jpg"))
```

**Review by adamjstewart — DISMISSED** [review]

File: torchgeo/datasets/millionaid.py [hunk 13]
```diff
@@ -0,0 +1,397 @@
...
+        "train_station",
+        "viaduct",
+        "wastewater_plant",
+        "wind_turbine",
+        "works",
+    ]
+
+    md5s = {
+        "train": "1b40503cafa9b0601653ca36cd788852",
+        "test": "https://eastus1-mediap.svc.ms/transform/zip?cs=fFNQTw",
```
Comment (calebrob6) [inline torchgeo/datasets/millionaid.py:180]: This isn't an MD5 hash

**Review by calebrob6 — DISMISSED** [review]

File: torchgeo/datasets/millionaid.py (hunk 13 above)
Comment (nilsleh) [reply torchgeo/datasets/millionaid.py:180]: Yes, the test set is the much larger file, and I am not able to download it completely, to compute the md5.

File: torchgeo/datasets/millionaid.py [hunk 14]
```diff
@@ -0,0 +1,397 @@
...
+        cls_label = [self.class_to_idx[label] for label in files["label"]]
+        label = torch.tensor(cls_label, dtype=torch.long)
+        sample = {"image": image, "label": label}
+
+        if self.transforms is not None:
+            sample = self.transforms(sample)
+
+        return sample
+
+    def _load_files(self, root: str) -> List[Dict[str, Any]]:
```
[the remark this answers is not in the document]
Comment (nilsleh) [reply torchgeo/datasets/millionaid.py:260]: I can only download the training set which is ~2GB large, and contains 10,000 image files. For this the `_load_files` function takes about one half of a second.

File: torchgeo/datasets/millionaid.py (hunk 13 above)
Comment (adamjstewart) [reply torchgeo/datasets/millionaid.py:180]: Can you try some kind of retry/continue on failure flag for wget/curl? Alternatively, you could email the authors and ask for the MD5 of the file (and warn them about the download issue).

Comment (nilsleh) [reply torchgeo/datasets/millionaid.py:180]: Sorry, what I meant to say is that my download issue is  that I do not have the disk space to download it completely.

Comment (adamjstewart) [reply torchgeo/datasets/millionaid.py:180]: @calebrob6 does your VM have space?

Comment (isaaccorley) [reply torchgeo/datasets/millionaid.py:180]: I'm downloading the test set to use for a side project and I'll report back with the hash.

File: torchgeo/datasets/millionaid.py [hunk 15]
```diff
@@ -0,0 +1,397 @@
...
+        "works",
+    ]
+
+    md5s = {
+        "train": "1b40503cafa9b0601653ca36cd788852",
+        "test": "https://eastus1-mediap.svc.ms/transform/zip?cs=fFNQTw",
+    }
+    url = {
+        "train": "https://eastus1-mediap.svc.ms/transform/zip?cs=fFNQTw",
+        "test": "https://eastus1-mediap.svc.ms/transform/zip?cs=fFNQTw",
```
Comment (isaaccorley) [inline torchgeo/datasets/millionaid.py:184]: Does this actually download the test set? These look like the same url to me. Also looks like the test set is made of multiple parts (e.g. `test.zip.001`, `test.zip.002`, etc.). Does `extract_archive` support this?

**Review by isaaccorley — CHANGES_REQUESTED** [review]
The test set is ~260GB and multiple parts compared to the train set which is a single zip file. I wonder if we even want users to be downloading this in a single sequential process. @adamjstewart

File: torchgeo/datasets/millionaid.py (hunk 15 above)
Comment (adamjstewart) [reply torchgeo/datasets/millionaid.py:184]: That is a good question, I'm not sure whether that will work or not. We've had to hack things to support deflate64-compressed zip files before, so multi-part zip files should be possible somehow.

File: torchgeo/datasets/millionaid.py (hunk 13 above)
Comment (isaaccorley) [reply torchgeo/datasets/millionaid.py:180]: The hash for `test.zip` is `51a63ee3eeb1351889eacff349a983d8`

File: torchgeo/datasets/millionaid.py (hunk 15 above)
Comment (nilsleh) [reply torchgeo/datasets/millionaid.py:184]: > Does this actually download the test set? These look like the same url to me.

If I look in the [OneDrive](https://whueducn-my.sharepoint.com/personal/longyang_whu_edu_cn/_layouts/15/onedrive.aspx?ga=1&id=%2Fpersonal%2Flongyang%5Fwhu%5Fedu%5Fcn%2FDocuments%2FMillion%2DAID%2FMillion%2DAID) and download the train or test folder, the download link for me are the same for some reason.

Comment (adamjstewart) [reply torchgeo/datasets/millionaid.py:184]: @isaaccorley you mentioned that you downloaded the test set and computed its MD5. Where did you download the test set from? Was it multiple parts when you downloaded and checksummed it?

Comment (adamjstewart) [reply torchgeo/datasets/millionaid.py:184]: @isaaccorley you mentioned that you downloaded the test set and computed its MD5. Where did you download the test set from? Was it multiple parts when you downloaded and checksummed it?

Comment (isaaccorley) [reply torchgeo/datasets/millionaid.py:184]: I used the OneDrive link above and right clicked on the `test` folder and it downloaded as `test.zip` which I then took the md5 of. The test set has multiple parts so even after you extract `test.zip` you still need to extract them together.

Comment (isaaccorley) [reply torchgeo/datasets/millionaid.py:184]: I actually just noticed that `test.zip` was corrupted and I couldn't unzip it completely so I think the hash may be incorrect. Going to try and download each of the test files individually.

Comment (calebrob6) [reply torchgeo/datasets/millionaid.py:184]: Was this resolved?

Comment (adamjstewart) [reply torchgeo/datasets/millionaid.py:184]: I don't think this was ever resolved, this is the only thing holding up this PR. If there's a single link for all data that's fine, we can download the data with one link, checksum it once, then extract it and extract and zip files that contains. I don't have the bandwidth/storage to download this myself, but can someone investigate this and see if it's even possible to download this dataset? If not, we could just remove the download logic until we figure it out.

Comment (adamjstewart) [reply torchgeo/datasets/millionaid.py:184]: Did anyone ever reach out to the dataset authors and see if we can rehost a single zip file on something like zenodo?

**Comment by adamjstewart** [comment]
Since the download logic doesn't currently work, I'm going to remove it so we can get this into 0.3.0. We can add download support in a future release.

File: docs/api/non_geo_datasets.csv [hunk 16]
```diff
@@ -16,6 +16,7 @@ Dataset,Task,Source,# Samples,# Classes,Size (px),Resolution (m),Bands
 `LandCover.ai`_,S,Aerial,"10,674",5,512x512,0.25--0.5,RGB
 `LEVIR-CD+`_,CD,Google Earth,985,2,"1,024x1,024",0.5,RGB
 `LoveDA`_,S,Google Earth,"5,987",7,"1,024x1,024",0.3,RGB
+`Million-AID`_,C,Google Earth,1M,51--73,,0.5--153,RGB
```
Comment (adamjstewart) [inline docs/api/non_geo_datasets.csv:19]: Does anyone know the range of image sizes?

**Review by adamjstewart — APPROVED** [review]
